### PR TITLE
Yarn transition work-in-progress

### PR DESCRIPTION
**Pull request by jywarren**
Solves #1215

### CODE REVIEW

**Comment by jywarren** [comment]
Running `bower-away` now

**Comment by jywarren** [comment]
Hmmm, ` Node version 0.10.29 is not supported, please use Node.js 4.0 or higher.`

**Comment by jywarren** [comment]
@ryzokuken any ideas?

**Comment by ryzokuken** [comment]
@jywarren who throws this error? Yarn?

If we're on Node 0.x, we should definitely migrate. Node 4 has reached EOL, and ideally, we should jump to atleast Node 8.

**Comment by jywarren** [comment]
Yes I tried to set a higher node version... But it didn't work! See the
issue I linked to above for yarn... Thanks!

**Comment by ryzokuken** [comment]
But travis allows you to use yarn. All you need to do is to check-in the yarn lockfile (which we should, and follow a policy similar to Gemfile.lock).

**Comment by jywarren** [comment]
Can you add a commit to this PR? Thanks!

**Comment by jywarren** [comment]
Trying to do more research in https://blog.travis-ci.com/2016-11-21-travis-ci-now-supports-yarn and add a `yarn.lock` -- though don't know what that has to do with the node version... For now just trying to set node version more explicitly in `.travis.yml` -- and failing!

**Comment by jywarren** [comment]
Hmm, looks like nvm is installing node 6 outside the container but not inside?

**Comment by jywarren** [comment]
maybe something like this? https://stackoverflow.com/questions/25899912/install-nvm-in-docker

**Comment by jywarren** [comment]
OK, locally i get:

```
[1/1] ⡀ phantomjs: Try to install PhantomJS globally
[-/1] ⡀ waiting...
[-/1] ⡀ waiting...
[-/1] ⡀ waiting...
error /home/warren/sites/plots2/node_modules/phantomjs: Command failed.
Exit code: 1
Command: node install.js
Arguments:
Directory: /home/warren/sites/plots2/node_modules/phantomjs
Output:
Considering PhantomJS found at /usr/bin/phantomjs
Found PhantomJS at /usr/bin/phantomjs ...verifying
PhantomJS detected, but wrong version 2.1.1 @ /usr/bin/phantomjs.
```

**Comment by icarito** [comment]
I would rather we used prepackaged node rather than nvm if possible there are debian packages we can use or ideally a ready made  base image. https://hub.docker.com/ is really slow for me but I'd look for an image with the right Node version and then install the ruby packages from the distro.

**Comment by icarito** [comment]
maybe we can switch to one of these: https://hub.docker.com/_/node/

**Comment by jywarren** [comment]
Hmm, interesting:

```
Your Gemfile.lock is corrupt. The following gem is missing from the DEPENDENCIES
section: 'hoe'
```

**Comment by jywarren** [comment]
Bundler version issue maybe? https://github.com/bundler/bundler/issues/4467

https://stackoverflow.com/questions/34276324/travis-reports-odd-message-of-corrupted-gemfile-lock

**Comment by icarito** [comment]
Thanks for enduring build experiments with me.
So in Debian Stretch we have specific versions that aren't going to change. We could aim at locking into those - this could have the impact of reducing some of our update burden because we can generally depend on Debian to support packages etc and security. Or we can ride the wave and try to keep up with newest.
I found there's a Docker image called [ruby-node](https://hub.docker.com/r/starefossen/ruby-node/) that tries that last path so I was tempted to try it, but the "maintained version" whatever that means would bump Ruby to 2.5.

Thinking over this again I think the best approach is to return to Ruby base image as that is what ruby-node's [Dockerfile](https://github.com/Starefossen/docker-ruby-node/blob/master/2-8/stretch/Dockerfile) does and install the desired Node version on top.

We really shouldn't think of nvm or rvm in containers as containers are disposable by nature and get rebuilt with the exact desired versions, generally no need for version managers except at build time.

**Comment by jywarren** [comment]
That's fine, i'm not wedded to `nvm` its just what we often use in development. I'm also fine with not being the latest node version. Honestly i'd just be happy to see the build succeed :-P it was a bit harder than I thought, given that the only issue remaining seems to be the node version!

**Comment by jywarren** [comment]
@icarito -- any suggestions on this?

**Comment by icarito** [comment]
Sooner rather than later we will need to upgrade our container to Stretch.
In this process we can add the `stretch-backports` repo that serves v8.11.2 LTS.
That would be the simplest path forward. The manual way of installing it looks a bit convulted:
https://github.com/Starefossen/docker-ruby-node/blob/master/2-8/stretch/Dockerfile

**Comment by jywarren** [comment]
> Sooner rather than later we will need to upgrade our container to Stretch.

OK, so let's try first opening a new PR for the Stretch port, then return to this. @ryzokuken if you can try this out, just be sure to link back to here so there's a next step.

**Comment by jywarren** [comment]
Seeing: 

```
Node version 0.10.29 is not supported, please use Node.js 4.0 or higher.
```

**Comment by ryzokuken** [comment]
@jywarren makes sense, it's bad if we're at Node v0.10.x, let's switch to Node v10.x instead.

**Comment by jywarren** [comment]
I'm fine with that! If you can move this forward at all, I'd be very
grateful. I'm just trying to move ahead with your excellent suggestion of
moving to Yarn, and am stuck here. Happy to accept help from anyone who can
get this to pass! Thanks!

**Comment by ryzokuken** [comment]
@jywarren I definitely should, sorry for procrastinating about this. I just keep getting bombarded with tasks. Let me look something up with week.

**Comment by jywarren** [comment]
no problem Ujjwal, i know you're busy! Just noting that I haven't been able
to make progress on this and could use help. Thanks!

**Comment by icarito** [comment]
Guys I've upgraded to Debian Stretch with https://github.com/publiclab/plots2/pull/2865 including Node 8. Merging that we can look at adding Yarn.

**Comment by jywarren** [comment]
Rebased!

**Comment by jywarren** [comment]
`/bin/sh: 1: npm: not found` -- well that would stop it from working 😉

**Comment by icarito** [comment]
Looks like the version of PhantomJS we install manually isn't compatible with newer node. 
We should probably switch to [phantomJS-prebuilt](https://www.npmjs.com/package/phantomjs-prebuilt).

```
error public/lib/phantomjs: Command failed.
Exit code: 1
Command: node install.js
Arguments: 
Directory: public/lib/phantomjs
Output:
module.js:549
    throw err;
    ^
Error: Cannot find module 'progress'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/app/public/lib/phantomjs/install.js:10:16)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

File: Dockerfile [hunk 1]
```diff
@@ -18,17 +18,19 @@ ENV PHANTOMJS_VERSION 2.1.1
 
 # Install dependencies
 RUN curl -sL https://deb.nodesource.com/setup_8.x | bash -
-RUN apt-get update -qq && apt-get install -y build-essential bundler libmariadbclient-dev ruby-rmagick libfreeimage3 wget curl procps cron make nodejs
+RUN apt-get update -qq && apt-get install -y bundler ruby-rmagick libfreeimage3 wget procps ruby-bundler ruby nodejs
 RUN wget https://github.com/Medium/phantomjs/releases/download/v$PHANTOMJS_VERSION/phantomjs-$PHANTOMJS_VERSION-linux-x86_64.tar.bz2 -O /tmp/phantomjs-$PHANTOMJS_VERSION-linux-x86_64.tar.bz2; tar -xvf /tmp/phantomjs-$PHANTOMJS_VERSION-linux-x86_64.tar.bz2 -C /opt ; cp /opt/phantomjs-$PHANTOMJS_VERSION-linux-x86_64/bin/* /usr/local/bin/
```
Comment (jywarren) [inline Dockerfile]: @icarito would you replace this line with `npm install phantomjs-prebuilt` ?

**Comment by jywarren** [comment]
hmm...

```

error public/lib/phantomjs: Command failed.
Exit code: 1
Command: node install.js
Arguments: 
Directory: public/lib/phantomjs
Output:
module.js:549
    throw err;
    ^
Error: Cannot find module 'bl'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
```

**Comment by jywarren** [comment]
@Souravirus -- once you're done with the Jasmine work, i wonder if you'd be able to take a look at this -- another key upgrade, although more related to Javascript...

I figure now you're deeper into the JavaScript stuff too! Thanks!!!

**Comment by Souravirus** [comment]
Yeah sure no problem

**Comment by jywarren** [comment]
Awesome, thank you! I just re-assigned it to you. Fantastic!

**Comment by jywarren** [comment]
Yeah i saw that too. @icarito thinks we can find a precompiled version of phantomjs, or switch the container and get this to compile... i mean, i think we have phantomjs running on our current master branch, so it's probably a matter of getting this all running in the right starting container?

**Comment by Souravirus** [comment]
But I am getting this phantomjs error in my local computer also. I don't know why

**Comment by Souravirus** [comment]
Let me find the reason behind it.

**Comment by Souravirus** [comment]
This is what I am getting when I am running `yarn --modules-folder public/lib` command.
error public/lib/phantomjs: Command failed.
Exit code: 1
Command: node install.js
Arguments: 
Directory: public/lib/phantomjs
Output:
module.js:549
    throw err;
    ^

Error: Cannot find module 'progress'
    at Function.Module._resolveFilename (module.js:547:15)
    at Function.Module._load (module.js:474:25)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/home/sourav/Documents/gsoc/plots2/public/lib/phantomjs/install.js:10:16)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)

**Comment by Souravirus** [comment]
I think some library is installing old version of phantomjs with it which is not being supported yarn. We have to find it and either remove it or find an alternative of it.

**Comment by Souravirus** [comment]
I tested this by removing all the libraries in public/lib folder and running `yarn --modules-folder public/lib` and then there is a problem with running install.js inside phantom.js.

**Comment by Souravirus** [comment]
Hmm by tracing that where is phantomjs is being installed by `yarn list` command. I got to know that it is being installed by ics dependency. I have to somehow make phantomjs-prebuilt installed instead of phantomjs.

**Comment by Souravirus** [comment]
I have ran the yarn install command by running it by ignoring the warnings by this command `yarn --ignore-engines --ignore-scripts --modules-folder public/lib && yarn postinstall` and then I am again running `yarn --modules-folder public/lib`. So,the dependency installation is now working fine. But there is one more error related to the bower which also we have to solve. 
Thanks!!

**Comment by jywarren** [comment]
`Error: ActionView::Template::Error: couldn't find file 'bootstrap/dist/js/bootstrap.min.js' with type 'application/javascript'`

I think this means that it cannot find these files... I remember with Bower we had to specify where it should be looking for them. And we customized it so Bower installed them in `/public/lib/` -- maybe we need to configure Yarn similarly?

**Comment by Souravirus** [comment]
Ok I was using webpacker to do this. But since the dependencies need to be installed in `public/lib`. I have to find some way about this fix.

**Comment by jywarren** [comment]
Hmm, where is webpack being used?

**Comment by Souravirus** [comment]
The dependency are installed in `node_modules` folder and then it is compiling and storing its files in `public/pack` folder.

**Comment by Souravirus** [comment]
I am sending a commit about the webpacker and lets see if it works.

**Comment by Souravirus** [comment]
Looks like I have to study a bit more about the webpacker.

**Comment by Souravirus** [comment]
Why the travis test is not running @jywarren ?

**Comment by Souravirus** [comment]
Yeah its working now.

**Comment by Souravirus** [comment]
Hi @jywarren, review the changes as this is working fine. Thanks!!

**Comment by jywarren** [comment]
Ah oops, looks like bower.json needs updating? Can you rebase and check that package.json is updated to match?

**Comment by icarito** [comment]
Production is built from Makefile `build` and `deploy-container` tasks.

We could use these make tasks in other CI scripts :-)

In the meantime we should adjust them from bower to yarn.

These are also used by Jenkins to for running tests and for staging.

**Comment by Souravirus** [comment]
Hi @icarito can you restart the gitlab ci. Looks like I don't have the permissions. Also an unknown error is coming:
`Job failed (system failure): Error response from daemon: Cannot link to a non running container: /runner-4e4528ca-project-7706596-concurrent-0-gitlab__dind-0 AS /runner-4e4528ca-project-7706596-concurrent-0-predefined-0/gitlab__dind (executor_docker.go:1014:0s`

**Comment by Souravirus** [comment]
Finally all tests pass @jywarren @icarito.

**Comment by Souravirus** [comment]
Hi @jywarren @icarito, pushing to unstable causes this error `No container found for mailman_1`. I have removed the bower install commands to which was not needed.

**Comment by Souravirus** [comment]
Here is the link https://jenkins.laboratoriopublico.org/job/Plots-Unstable/235/console of the error. If possible please see this.

**Comment by Souravirus** [comment]
Hi @jywarren @icarito yarn is running fine in jenkins but `docker-compose exec -T mailman bash -c "echo 172.19.0.1 smtp >> /etc/hosts"` command is giving this error `No container found for mailman_1`. Can anyone help me with this? I have no idea about mailman. Thank you!!

**Comment by icarito** [comment]
Hi!
I've checked at unstable the logs for mailman and they are showing this error to start:
```
icarito@tycho:/srv/plots_unstable/plots2$ docker-compose -f containers/docker-compose-unstable.yml logs mailman
WARNING: The RAILS_ENV variable is not set. Defaulting to a blank string.
WARNING: The SECRET_KEY_BASE variable is not set. Defaulting to a blank string.
WARNING: The RECAPTCHA_SECRET_KEY variable is not set. Defaulting to a blank string.
WARNING: The RECAPTCHA_SITE_KEY variable is not set. Defaulting to a blank string.
WARNING: The SERVER_ADDRESS variable is not set. Defaulting to a blank string.
WARNING: The USERNAME variable is not set. Defaulting to a blank string.
WARNING: The EMAIL_PASSWORD variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_GOOGLE_APP_KEY variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_GOOGLE_APP_SECRET variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_GITHUB_APP_KEY variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_GITHUB_APP_SECRET variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_TWITTER_APP_KEY variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_TWITTER_APP_SECRET variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_FACEBOOK_APP_KEY variable is not set. Defaulting to a blank string.
WARNING: The OAUTH_FACEBOOK_APP_SECRET variable is not set. Defaulting to a blank string.
Attaching to containers_mailman_1
mailman_1  | /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/spec_set.rb:87:in `block in materialize': Could not find rack-test-1.0.0 in any of the sources (Bundler::GemNotFound)
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/spec_set.rb:81:in `map!'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/spec_set.rb:81:in `materialize'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/definition.rb:159:in `specs'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/definition.rb:218:in `specs_for'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/definition.rb:207:in `requested_specs'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/runtime.rb:109:in `block in definition_method'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/runtime.rb:21:in `setup'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler.rb:101:in `setup'
mailman_1  |    from /usr/local/lib/ruby/gems/2.4.0/gems/bundler-1.15.4/lib/bundler/setup.rb:19:in `<top (required)>'
mailman_1  |    from /usr/local/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:133:in `require'
mailman_1  |    from /usr/local/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:133:in `rescue in require'
mailman_1  |    from /usr/local/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:40:in `require'
mailman_1  |    from script/mailman_server:3:in `<main>'
```

**Comment by icarito** [comment]
This is why it's failing to add the host IP - mailman needs to be running!

**Comment by icarito** [comment]
Thanks for your hard work!

**Comment by Souravirus** [comment]
Ah! ok, so I guess I need to add rack-test gem. Let's see what happens after adding it. Thanks for your help @icarito

**Comment by Souravirus** [comment]
The error `No container found for mailman_1` still persists. Can you provide the jenkins log again? Also you can see my changes in Makefile now. Please review it  once @icarito.

**Comment by jywarren** [comment]
Just mentioning that if we don't get to this earlier, @icarito and I will check in about it this friday at our meeting and try to get this to pass and merge! Thanks @Souravirus !!!

**Comment by Souravirus** [comment]
Yeah it would be great @jywarren, I would really like to help with this. I am just unable to deploy it to the `unstable` branch. It would remove all the merge conflicts in this. Thanks

**Comment by jywarren** [comment]
Hi @Souravirus, @icarito has updated the setup and `unstable` should be working now, can you try again?

**Comment by jywarren** [comment]
OK figure out it was ruby 2.4.1, but we're now ruby 2.4.4 -- changed it and let's see! It should be good to push to `unstable` now too.

**Comment by Souravirus** [comment]
ok lets see if it passes the test here. Meanwhile, I am pushing this to unstable too.

**Comment by jywarren** [comment]
oops i just did too!

File: Makefile [hunk 2]
```diff
@@ -58,5 +57,3 @@ setup-complete:
 	rvm install ruby-2.1.2
```
Comment (jywarren) [inline Makefile]: do we need to change this too?

**Comment by jywarren** [comment]
Also i see: `Could not find i18n-1.0.1 in any of the sources` -- something needed there too?

File: Makefile (hunk 2 above)
Comment (Souravirus) [reply Makefile]: Yeah, we have to

**Comment by jywarren** [comment]
want to go ahead and modify that?

**Comment by Souravirus** [comment]
Yeah I will do it certainly but currently I am having problem with updating the ruby version to 2.4.4. I am trying to fix it.

File: Dockerfile [hunk 3]
```diff
@@ -20,15 +20,19 @@ ENV PHANTOMJS_VERSION 2.1.1
...
+# Install yarn
+RUN npm config set strict-ssl false
+RUN npm install -g yarn
 
 # Install bundle of gems
 WORKDIR /tmp
 ADD Gemfile /tmp/Gemfile
-ADD Gemfile.lock /tmp/Gemfile.lock
-RUN bundle install --jobs 4
+#ADD Gemfile.lock /tmp/Gemfile.lock  # *DISABLED TO TEST IMAGE SWITCH*
```
Comment (icarito) [inline Dockerfile]: This is likely the problem! `mini_mime` is only declared in `Gemfile.lock`!
Hope it helps!

Comment (Souravirus) [reply Dockerfile]: Ok lets see if it works by removing this comment

**Comment by Souravirus** [comment]
Hi @jywarren, @icarito now all the tests are passing here as well as the deployment to `unstable` branch is working fine. Please review this.

**Comment by jywarren** [comment]
This is great. 🎉  Can you give us a checklist on anything that has to be done on production for this to work? Or do you and @icarito anticipate that it'll run smoothly. 

And one more thing - have you been able to double check that the most current bower.json file has been copied over (with accurate version numbers) to package.json?

**Comment by jywarren** [comment]
super, awesome.

File: .travis.yml [hunk 4]
```diff
@@ -1,5 +1,12 @@
 sudo: required
 
+# these didn't work:
+#before_script: # install v6.x of nodejs
+#- nvm install 6
+#language: node_js
+#node_js: '6'
+#cache: yarn # This caches `$HOME/.yarn-cache`
+
```
Comment (icarito) [inline .travis.yml]: Please remove these comments unless they help somehow!

File: Makefile [hunk 5]
```diff
@@ -7,13 +7,12 @@ build:
...
 	docker-compose build --pull
 	docker-compose exec web rake db:migrate
-	docker-compose exec web bower install --allow-root
-	docker-compose exec web bower update --allow-root
 	docker-compose exec web rake assets:precompile
 	docker-compose exec web rake tmp:cache:clear
 	docker-compose down --remove-orphans
 	rm -f ./tmp/pids/server.pid
 	docker-compose up -d
+	docker-compose exec -T web yarn --ignore-engines --ignore-scripts --modules-folder public/lib
```
Comment (icarito) [inline Makefile]: You've removed the javascript package installation from line 10/11 and put it into line 15. This effectively waits until *after* rebooting the Rails app to get going with Javascript installation. So if Javascript installation fails, you've already restarted the app possibly broken by missing Javascript dependencies. If you move this line to line 10, then if yarn fails the app is still running fine and no harm was done. If you agree please do so! Thanks!

File: Makefile [hunk 6]
```diff
@@ -23,11 +22,10 @@ redeploy-container:
 deploy-container:
 	docker-compose run web sleep 5
 	docker-compose run web rake db:migrate
-	docker-compose run web bower install --allow-root
-	docker-compose run web bower update --allow-root
 	docker-compose run web rake assets:precompile
 	rm -f ./tmp/pids/server.pid
 	docker-compose up -d
+	docker-compose exec -T web yarn --ignore-engines --ignore-scripts --modules-folder public/lib
```
Comment (icarito) [inline Makefile]: Again here, you wait till after deploying the container to install Javascript packages, I think the same applies here. Thanks!

**Comment by icarito** [comment]
@jywarren This morning it came to me that you may find it won't deploy cleanly with `redeploy-container` method (because in the first part of the process it'll exec yarn but yarn isn't installed in the current image). I guess this is why @Souravirus put this command *after* restarting the container (because then `yarn` will be available).
In these cases (like when we upgrade ruby) - when `redeploy-container` fails, our Makefile has `deploy-container` task which we can use to deploy from a stopped container status.

File: .travis.yml (hunk 4 above)
Comment (Souravirus) [reply .travis.yml]: Ok I wiil remove these

File: Makefile (hunk 5 above)
Comment (Souravirus) [reply Makefile]: Ok I will move the yarn installation command to line 11. Thanks

**Comment by Souravirus** [comment]
Yeah right @icarito putting yarn before restarting the container is giving error as you can see in https://jenkins.laboratoriopublico.org/job/Plots-Unstable/296/console . So, I guess best is to use the yarn command after starting the container. What do you think @icarito.

**Comment by icarito** [comment]
I think this is working fine in unstable now! @jywarren and the docker-compose bit should make it reliable to upgrade onwards.

**Comment by jywarren** [comment]
OK, shall I merge this in then? SUPER!!!

**Comment by Souravirus** [comment]
Yeah its ready @jywarren

**Comment by jywarren** [comment]
🎉 !!!

**Comment by jywarren** [comment]
This is live!!!!!

**Comment by jywarren** [comment]
This is live!!!!!

**Comment by Souravirus** [comment]
We have to change the documentation for the change from bower to yarn. I guess making a first-timer issue of it will work.

**Comment by Souravirus** [comment]
I guess we can make it a code-in task too. What do you think @jywarren

**Comment by jywarren** [comment]
That'd be great!!!

**Pull request merged**